lib/pbio/drv/usb/usb_stm32.c

```diff
@@ -164,11 +304,65 @@ PROCESS_THREAD(pbdrv_usb_process, ev, data) {
 
     PROCESS_BEGIN();
 
+    // Prepare to receive the first packet
+    USBD_Pybricks_ReceivePacket(&husbd);
+
+    etimer_set(&timer, 500);
+
     for (;;) {
         PROCESS_WAIT_EVENT();
 
         if (bcd_busy) {
             bcd_busy = PT_SCHEDULE(pbdrv_usb_stm32_bcd_detect(&bcd_pt));
+
+            if (bcd_busy) {
+                continue;
+            }
+        }
+
+        if (pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN_REQUEST) ||
+            pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN)) {
+            if (pbio_oneshot(true, &pwrdn_oneshot)) {
+                USBD_Stop(&husbd);
+            }
+
+            continue;
+        }
+
+        if (usb_in_sz) {
+            if ((usb_in_sz >= UUID_SZ) &&
+                pbio_uuid128_le_compare(usb_in_buf, pbio_pybricks_command_event_char_uuid)) {
+                pbsys_command(usb_in_buf + UUID_SZ, usb_in_sz - UUID_SZ);
+            }
+
+            usb_in_sz = 0;
+
+            // Prepare to receive the next packet
+            USBD_Pybricks_ReceivePacket(&husbd);
+        }
+
+        if (usb_out_sz) {
+            continue;
+        }
+
+        new_status_flags = pbsys_status_get_flags();
+
+        if (usb_stdout_sz) {
```

Comment:
Like the ring buffer?

What sort of responses do we expect to send? This (and the pybricksdev changes) are currently not doing anything with responses and it's working so I wasn't sure if it was necessary.

Reply:
> Like the ring buffer?

No. Rather like a linked list that has at most 3 nodes, one for each message type (stdout, status, write response).

Reply:
> What sort of responses do we expect to send?

The status code returned from `pbsys_command()`.

> This (and the pybricksdev changes) are currently not doing anything with responses and it's working so I wasn't sure if it was necessary.

As long as you are sending correct data at the correct time, you would never get an error, but it would be better if we passed along errors in case someone ever tries to do something not allowed.

Reply:
> Rather like a linked list that has at most 3 nodes

Or 2 nodes, one for each `pbio_pybricks_event_t` and put the write responses on a separate endpoint.

Reply:
I think using separate endpoints will complicate things because you have to use more functionality in the STM32 USB driver. This is harder, especially if you don't have JTAG. I think the approach we have here using UUID to multiplex messages over a single endpoint is easier. What are your thoughts?

Reply:
I'm not sure I understand what makes it more complicated. From looking at the code, all endpoints (other than 0) are handled the same. It seems like it is just a matter of tweaking the config a bit and adding a few if statements to the existing handlers.

Reply:
If we do leave it as one endpoint, I would rather use a single byte instead of a 16-byte UUID. The UUID is taking up 25% of the available 64 bytes in each packet.

Reply:
I looked into multiple endpoints and it involves a lot of refactoring. What do you think about adding another event (value 2) to capture the command response.

Reply:
Yeah, I agree on the UUID not being necessary. I have a better understanding of the protocol now so that makes sense.

What I was proposing was a little flatter in that all hub-to-host messages are events and all host-to-hub messages are commands. We would add a new event type 0x02 to cover command responses. So like this:

Hub to host endpoint:
```
0x00 = STOP_USER_PROGRAM
0x01 = START_USER_PROGRAM
...
0x06 = WRITE_STDIN
```

Host to hub endpoint:
```
0x00 = STATUS_REPORT
0x01 = WRITE_STDOUT
0x02 = COMMAND_RESPONSE <- new, not needed by BLE
```
Writing "Hello World" to stdout would look something like `01 48 65 6C 6C 6F 20 57 6F 72 6C 64`

I think what you were proposing is a little more hierarchical, which is fine, but I'm not sure it is necessary. To illustrate:

Hub to host endpoint:
```
0x00 = Write response
0x01 = Pybricks event
     0x00 = STATUS_REPORT
     0x01 = WRITE_STDOUT
```
Writing "Hello World" to stdout would look something like `01 01 48 65 6C 6C 6F 20 57 6F 72 6C 64`

Host to hub endpoint:
```
0x00 = Pybricks command
     0x00 = STOP_USER_PROGRAM
     0x01 = START_USER_PROGRAM
     ...
     0x06 = WRITE_STDIN
```
What's your preference?

Reply:
I would rather not extend the Pybricks protocol with things that only apply to USB, so I would prefer the latter more hierarchical approach. This will also make it easier to add, e.g. the Nordic UART stuff over USB later if we decide we need it.